accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1248,13 +1248,122 @@ the `LEGACY_ERASURE` of the types.
+ignoring legacy type annotations as described above.  Hence, an instance of
+`C<T0>` compares identical to `C<T1>` if `T0` and `T1` have the same normal form
+(up to the identity of bound variables), and the objects are otherwise
+identical.
+
+Implementations of the Dart runtime semantics rely on canonicalization of
+constant objects to allow the identity semantics specified above to be
+implemented as fast pointer equality checks on the reference to the canonical
+form.  The definition above defines equivalence classes of constant objects for
+which we must choose the canonical representative.  The choice of this
+representative is observable in mixed mode programs, since instances with
+different degrees of "legacy-ness" in their type arguments are considered
+identical, but may contain operations which perform casts and instance checks
+which will evaluate differently depending on whether a legacy type or a
+non-legacy type is used in the canonical representative.  For example:
+
+```dart
+// null safe code.
+class C<T> {
+  void test(Object? o) {
+    assert(o is T);
+  }
+  const C(Object? o) : assert(o is T);
+}
+
+// If the canonical instance uses `int`, this is a compile time error
+// If the canonical instance uses `int*`, this is not a compile time error
+const c1 = C<int>(null);
+
+// If the canonical instance uses `int`, this throws
+// If the canonical instance uses `int*`, this does not throw
+void test1() => c1.test(null);
+
+
+// Opted out code
+
+// If the canonical instance uses `int`, this is a compile time error
+// If the canonical instance uses `int*`, this is not a compile time error
+const c2 = C<int>(null);
+
+// If the canonical instance uses `int`, this throws
+// If the canonical instance uses `int*`, this does not throw
+void test1() => c2.test(null);
+```
+
+We therefore define the choice of the canonical instance representing an
+equivalence class of constant objects as follows.
+
+With sound null checking, all generic const constructors and generic const
+literals are evaluated using the type arguments provided, and canonicalization
+is performed with respect to the normal form of the type arguments.  This
+ensures that with sound null checking, the final consistent semantics are
+obeyed, since it is not observable which instance is chosen as the canonical
+representative in sound mode.
+
+With unsound null checking, all generic const constructors and generic const
+literals are additionally treated as if all type arguments passed to them were
+legacy types regardless of whether the constructed class was defined in a legacy
+library or not, and regardless of whether the constructor invocation or literal
+occured in a legacy library or not.  Specifically, a constructor invocation or
+object literal with generic type parameters `Ti` is treated as if the parameters
+were **LEGACY_TYPE**('Ti') as defined below. Implementations may choose to
+eagerly normalize the type arguments before applying the legacy rewrite, as
```

Comment:
I believe normalization here is intended to be an optimization opportunity, and it should not be observable because it only affects the creation of a (canonical) type object, and there is no way (other than `.toString()`) to detect whether normalization occurred (because normalization is used for `Type.==`, and `Type.==` is used to determine canonicalization).

It would then be helpful to make this sentence commentary, and mention that it is not observable: It would not make any difference if this sentence were omitted, so it shouldn't be normative text.